### PR TITLE
kube-apiserver: drain HTTP/2 early during termination

### PR DESCRIPTION
**Pull request by sttts**
Picking upstream https://github.com/kubernetes/kubernetes/pull/92136.

### CODE REVIEW

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go [hunk 1]
```diff
@@ -47,6 +47,9 @@ func WithProbabilisticGoaway(inner http.Handler, chance float64) http.Handler {
 		decider: &probabilisticGoawayDecider{
 			chance: chance,
 			next: func() float64 {
+				if isTerminating() {
```
Comment (p0lyn0mial) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go:50]: nice !

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go [hunk 2]
```diff
@@ -56,13 +59,25 @@ func WithProbabilisticGoaway(inner http.Handler, chance float64) http.Handler {
...
+
 // goaway send a GOAWAY to client according to decider for HTTP2 requests
 type goaway struct {
 	handler http.Handler
 	decider GoawayDecider
 }
 
-// ServeHTTP implement HTTP handler
+// ServeHTTP implements HTTP handler
 func (h *goaway) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```
Comment (p0lyn0mial) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go:81]: this is on HTTP level so it should work

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go [hunk 3]
```diff
@@ -56,13 +59,25 @@ func WithProbabilisticGoaway(inner http.Handler, chance float64) http.Handler {
 	}
 }
 
+// WithTerminationGoaway sends a GOAWAY when isTerminating is returning true.
+func WithTerminationGoaway(inner http.Handler, isTerminating func() bool) http.Handler {
```
Comment (p0lyn0mial) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go:63]: nice !

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go [hunk 4]
```diff
@@ -39,6 +39,7 @@ type ServerRunOptions struct {
 	MaxMutatingRequestsInFlight int
 	RequestTimeout              time.Duration
 	GoawayChance                float64
+	TerminationGoaway           bool
```
Comment (p0lyn0mial) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go:42]: I haven't seen the usage of this variable in your code.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go [hunk 5]
```diff
@@ -194,6 +196,9 @@ func (s *ServerRunOptions) AddUniversalFlags(fs *pflag.FlagSet) {
 		"This argument sets the fraction of requests that will be sent a GOAWAY. Clusters with single apiservers, or which don't use a load balancer, should NOT enable this. "+
 		"Min is 0 (off), Max is .02 (1/50 requests); .001 (1/1000) is a recommended starting point.")
 
+	fs.BoolVar(&s.TerminationGoaway, "termination-goaway", s.TerminationGoaway, ""+
```
Comment (p0lyn0mial) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go:199]: why do we want to export it via flag? what will be the default value?

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go (hunk 4 above)
Comment (sttts) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go:42]: fixed.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go (hunk 5 above)
Comment (sttts) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go:199]: false, i.e. old behaviour.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go [hunk 6]
```diff
@@ -56,13 +56,25 @@ func WithProbabilisticGoaway(inner http.Handler, chance float64) http.Handler {
...
+	return &goaway{
+		handler: inner,
+		decider: GoawayDeciderFunc(
+			func(r *http.Request) bool {
+				return isTerminating()
```
Comment (p0lyn0mial) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go:65]: maybe we should log these requests? they would end up in the new termination.log file.

Comment (sttts) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go:65]: we log all on termination anyway.

**Pull request merged**